packages/rule-tester/src/RuleTester.ts

```diff
@@ -929,130 +940,161 @@ export class RuleTester extends TestFramework {
             );
           }
 
+          assert.ok(
+            !message.suggestions || hasOwnProperty(error, 'suggestions'),
+            `Error at index ${i} has suggestions. Please specify 'suggestions' property on the test error object.`,
+          );
           if (hasOwnProperty(error, 'suggestions')) {
             // Support asserting there are no suggestions
-            if (
-              !error.suggestions ||
-              (isReadonlyArray(error.suggestions) &&
-                error.suggestions.length === 0)
-            ) {
-              if (
-                Array.isArray(message.suggestions) &&
-                message.suggestions.length > 0
-              ) {
-                assert.fail(
-                  `Error should have no suggestions on error with message: "${message.message}"`,
-                );
-              }
-            } else {
-              assert(
-                Array.isArray(message.suggestions),
-                `Error should have an array of suggestions. Instead received "${String(
-                  message.suggestions,
-                )}" on error with message: "${message.message}"`,
+            const expectsSuggestions = Array.isArray(error.suggestions)
+              ? error.suggestions.length > 0
+              : Boolean(error.suggestions);
+            const hasSuggestions = message.suggestions !== void 0;
+            // @ts-expect-error -- we purposely don't verify for undefined
+            const messageSuggestions: Linter.LintSuggestion[] =
+              message.suggestions;
+
+            if (!hasSuggestions && expectsSuggestions) {
+              assert.ok(
+                !error.suggestions,
+                `Error should have suggestions on error with message: "${message.message}"`,
```

Comment:
```suggestion
            if (message.suggestions === void 0 && expectsSuggestions) {
              assert.ok(
                !error.suggestions,
                `Error should have suggestions on error with message: "${message.message}"`,
```

[Refactor] `hasSuggestions` is only used twice, perhaps we can inline a `message.suggestions !== void 0` check in the if conditions, that way TypeScript will understand that it is safe to use `message.suggestions` without putting `// @ts-expect-error`?